### ISSUE TITLE
QGTOnTheFly chunked solve returns x_ref / n_devices when chunk_size < samples_per_shard (with pdf)

### ISSUE THREAD

**Issue by gcarleo** [netket/netket#2228]
## Summary

Under `NETKET_EXPERIMENTAL_SHARDING_CPU > 1`, solving a weighted `QGTOnTheFly` with `chunk_size < samples_per_shard` (where `samples_per_shard = n_samples // jax.device_count()`) returns a result that is **exactly** `x_ref / n_devices` instead of `x_ref`, silently corrupting any downstream SR update. Setting `chunk_size >= samples_per_shard` or `chunk_size=None` gives the correct answer.

The factor is exactly `1 / n_devices` for every `chunk_size` below the per-shard size, which strongly suggests a missing shard-axis reduction in the chunked multi-chunk-per-shard code path when a `pdf` is provided.

## Versions

- netket `3.22.0.dev3`
- jax `0.10.0`
- flax `0.12.7`
- Python `3.13.5`
- macOS (Darwin 23.6.0), CPU backend

## Minimal reproducer

Standalone, uses only stock NetKet primitives:

```python
"""NETKET_EXPERIMENTAL_SHARDING_CPU=4 python mre_qgt_chunk.py"""
from __future__ import annotations

import os
os.environ.setdefault("NETKET_EXPERIMENTAL_SHARDING_CPU", "4")

import jax
import jax.numpy as jnp
import numpy as np
import netket as nk
from jax.flatten_util import ravel_pytree
from netket.optimizer.qgt.qgt_onthefly import QGTOnTheFly_DefaultConstructor
from netket.optimizer.solver import pinv_smooth


def solve_qgt(vstate, samples, pdf, rhs, chunk_size):
    qgt = QGTOnTheFly_DefaultConstructor(
        vstate._apply_fun, vstate.parameters, vstate.model_state,
        samples, pdf=pdf, holomorphic=True, chunk_size=chunk_size,
    )
    x, _ = qgt.solve(pinv_smooth(rcond=1e-8, rcond_smooth=1e-8), rhs)
    x_flat, _ = ravel_pytree(x)
    return x_flat


def main():
    n_sites = 8
    n_samples = 4096
    alpha = 2
    seed = 0

    hi = nk.hilbert.Spin(s=0.5, N=n_sites)
    model = nk.models.RBM(alpha=alpha, param_dtype=complex)

    sampler = nk.sampler.MetropolisSampler(
        hi, rule=nk.sampler.rules.LocalRule(),
        n_chains=max(16, jax.device_count() * 4),
    )
    vstate = nk.vqs.MCState(sampler, model, n_samples=n_samples, seed=seed)

    samples = jnp.asarray(
        np.asarray(vstate.samples).reshape(-1, n_sites)
    )
    rng = np.random.default_rng(seed)
    omega = rng.uniform(0.1, 1.0, size=samples.shape[0])
    pdf = jnp.asarray(omega / omega.sum())

    params_flat, unravel = ravel_pytree(vstate.parameters)
    rhs_flat = rng.normal(size=params_flat.size) + 1j * rng.normal(
        size=params_flat.size
    )
    rhs = unravel(jnp.asarray(rhs_flat))

    n_dev = jax.device_count()
    per_shard = n_samples // n_dev
    print(f"jax.device_count()={n_dev}, n_samples={n_samples}, "
          f"samples_per_shard={per_shard}")
    print(f"{'chunk_size':>14} {'max|Δ|':>14} {'|x_cs|/|x_ref|':>18}")
    print("-" * 55)

    x_ref = solve_qgt(vstate, samples, pdf, rhs, None)
    for cs in [None, per_shard * 2, per_shard, per_shard // 2,
               per_shard // 4, 128, 64]:
        if cs is None:
            x = x_ref
        else:
            x = solve_qgt(vstate, samples, pdf, rhs, cs)
        diff = float(jnp.max(jnp.abs(x - x_ref)))
        ratio = float(jnp.linalg.norm(x) / jnp.linalg.norm(x_ref))
        print(f"{str(cs):>14} {diff:>14.3e} {ratio:>18.5f}")


if __name__ == "__main__":
    main()
```

## Observed output

```
jax.device_count()=4, n_samples=4096, samples_per_shard=1024
    chunk_size         max|Δ|     |x_cs|/|x_ref|
-------------------------------------------------------
          None      0.000e+00            1.00000
          2048      0.000e+00            1.00000
          1024      0.000e+00            1.00000
           512      7.986e+02            0.25000
           256      7.986e+02            0.25000
           128      7.986e+02            0.25000
            64      7.986e+02            0.25000
```

Every row with `chunk_size < 1024` gives `|x_cs| / |x_ref| = 0.25 = 1 / n_devices`, and all such rows agree on the wrong vector (same `max|Δ|`), so the output is a deterministic rescaling of the correct result — not a stochastic error.

## Expected behavior

All rows should agree to ~1e-12 regardless of `chunk_size`, as is the case when `chunk_size >= samples_per_shard`.

## Likely source

`netket/optimizer/qgt/qgt_onthefly_logic.py::_Odagger_DeltaO_v` (~L135–145):

```python
def _Odagger_DeltaO_v(forward_fn, params, model_state, samples, v, chunk_size, pdf=None):
    w = _O_jvp(forward_fn, params, model_state, samples, v, chunk_size)
    if pdf is None:
        w = w * (1.0 / samples.shape[0])
        w = w - jnp.mean(w)  # w/ JAX sharding
    else:
        w = pdf * (w - pdf @ w)
    res = _OH_w(forward_fn, params, model_state, samples, w, chunk_size)
    return res
```

`_O_jvp` and `_O_vjp` are each wrapped with `sharding_decorator(..., reduction_op_tree=jax.lax.psum)` (for `_O_vjp`) or plain `sharding_decorator` (for `_O_jvp`), so inside each you see the per-shard slice of `samples` and `v`. The `scanmap`+`chunk(samples, chunk_size)` inside those functions splits the per-shard slice into chunks; `scan_append` and `scan_reduce` accumulate over those chunks (per-shard), and the outer `sharding_decorator` is responsible for stitching the shards back together.

At `chunk_size >= samples_per_shard`, `chunk()` produces exactly one chunk per shard and the accumulation reduces to a no-op — the path goes through what is effectively the "1 chunk" case and the outer `psum`/reshape combines shards correctly.

At `chunk_size < samples_per_shard`, the per-shard scan produces multiple partial results that get accumulated via `scan_append`/`scan_reduce` **before** the decorator's shard reduction runs, and the cross-shard normalization/`pdf @ w` step (outside the `@sharding_decorator` wrappers) sees a `w` whose scaling has already been divided by `n_devices` but hasn't been `psum`ed. The precise factor of `1 / n_devices` in the reproducer matches exactly this: the inner `_O_jvp` result is implicitly an average per shard rather than a sum, so `pdf @ w` produces the wrong normalization and every subsequent step carries the `1/n_devices` factor.

Suggested repro fix to investigate: make sure the `pdf * (w - pdf @ w)` expression in `_Odagger_DeltaO_v` is evaluated on a correctly shard-reduced `w`, either by wrapping the whole `_Odagger_DeltaO_v` in a `sharding_decorator` (with appropriate `reduction_op_tree`) or by adjusting the chunked inner reduction so that its output is shard-summed before the `pdf`-weighted step. The non-`pdf` branch relies on `jnp.mean(w)` (which JAX + NamedSharding handles correctly), so it doesn't exhibit the bug — but the `pdf` branch's manual `pdf @ w` has to be either explicitly `psum`-reduced or executed inside the same shard-map as the rest.

## Workaround

`chunk_size = None` or `chunk_size >= n_samples // jax.device_count()` (i.e. at most one chunk per shard). In practice this means chunking inside a shard is currently unusable when a nontrivial `pdf` is passed.
